README.md
# PREVIEW TO v2.0.0

Currently i work hard on getting out v2.0.0 of i18next. It's a complete rewrite of the current code base and will assert i18next is ready for the next things that will come.

Current state is still alpha - but i highly encourage maintainers of modules using i18next - to take the early chance to play with the upcoming version and starting to migrate to v2.

For people using i18next in production i recommend waiting for first beta version to play around.

more info could be found on: [Migration Guide](http://i18next.github.io/i18next.com/docs/migration/)

# Introduction

[![Build Status](https://secure.travis-ci.org/i18next/i18next-node.png)](http://travis-ci.org/i18next/i18next-node)

Project goal is to provide the same easy way to translate a website on serverside like in 
[i18next](https://github.com/i18next/i18next) on the clientside:

- Translation inside your server code or template
- loading resourcefiles
- update resourcefiles with missing strings
- interpolation, plurals, nesting or resources
- serve the same resources to the clientside

# installation

	npm install i18next

# usage

First require the module and init i18next:

```js
var i18next = require('i18next');
i18next.init(); // for options see i18next-node gh-page
```

Register the express middleware, so we can check current language settings:

```js
// Configuration
app.configure(function() {
    app.use(express.bodyParser());
    app.use(i18next.handle);
    app.use(app.router);

    [...]
});
```

Register AppHelper so you can use the translate function in your template:

```js
i18next.registerAppHelper(app)
```

Now you can (depending on your template language) do something like this in your template(jade example):

```jade
body
span= t('app.name')
```

To serve the clientside script and needed routes for resources and missing keys:

```js
i18next.serveClientScript(app)
    .serveDynamicResources(app)
    .serveMissingKeyRoute(app);
```

now you can add the script to you page and use i18next on the client like on the server:

```jade
script(src='i18next/i18next.js', type='text/javascript')
```

```js
$.i18n.init([options], function() { 
    $('#appname').text($.t('app.name'));
});
```

for more information on clientside usage have a look at [i18next](http://i18next.github.com/i18next/)

# sample

- [i18next-node_Sample](https://github.com/i18next/i18next-node/tree/master/sample)

# License

Copyright (c) 2014 Jan Mühlemann

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.


CHANGELOG.md
### v1.10.6
- update client
- Fix ensureIsSupportedLanguage middleware injection. [PR 200](https://github.com/i18next/i18next-node/pull/200)

### v1.10.5
- [SECURITY] update client to include - merges Reimplement XSS-vulnerable sequential replacement code [PR 443](https://github.com/i18next/i18next/pull/443)

### v1.10.4
- fixes appending lng in resStore if a namespace is omitted

### v1.10.3
- allows backends to promote an error having err.i18nSkipOnError that will prevent setting an empty resource set and allows a reload on next access this way (usable in remotesync if remote in unavailable)

### v1.10.2
- update client
- fixes fetchOne issue overwriting already loaded resources

### v1.10.1
- fixes client version

### v1.10.0
- [BREAKING] new callbacks will be node.js conform function(err, t) | Forward the error from sync fetch methods to the init callback function [PR 402](https://github.com/i18next/i18next/pull/402)
- update client
- Stop unnecessary reloading all namespaces [PR 196](https://github.com/i18next/i18next-node/pull/196)

### v1.9.0
- updates i18next-client to 1.9.0
- fixes issue with cache control header [issue 169](https://github.com/i18next/i18next-node/issues/169)
- Ensure language support in addRoute made optional [PR 174](https://github.com/i18next/i18next-node/pull/174)
- Removed pluralExtensions lookup from detectLanguageFromPath [PR 181](https://github.com/i18next/i18next-node/pull/181)
- solves issue: method 'exists' not available in req.i18n [issue 187](https://github.com/i18next/i18next-node/issues/187)

### v1.7.10
- sync.load refactor to reload not yet loaded namespaces - not break on loaded language

### v1.7.9
- fixes issue with saveMissing [issue 165](https://github.com/i18next/i18next-node/issues/165)

### v1.7.8
- update client to 1.7.7

### v1.7.6
- updated client to 1.7.6
- multiple handlers for addRoute
- options.detectLanguageFn(req, res) --> allows on detection Function (eg. from storage)
- allow cache=false maxAge=0
- JSON5 support
- strip BOM
- try JSON.parse

### v1.7.3
- updated client to 1.7.3
- fixes issue with using WT with filesync [issue 116](https://github.com/jamuhl/i18next-node/issues/116)
- fixes issue with detectLngFromPath if route level [pullrequest 121](https://github.com/jamuhl/i18next-node/pull/121)

### v1.7.1
- update client for some typos

### v1.7.0
- setLng via req.i18n.setLng (same as setting req.lng)
- support cookie domain (options cookieDomain = '')
- updated client

### v1.6.8
- fixes issue with saveMissing key/value - it should not throw error on unknown language namespace combination
- fixes issue with fallbackLng = false in combination with sendMissingTo = fallback
- updated client

### v1.6.7
- updated client

### v1.6.6
- fixing typo issue in client

### v1.6.5
- updated client
- set json intend via init options jsonIndent = integer

### v1.6.4
- fixes concurrancy issue for express version < 3

### v1.6.3
- update with pre client
- fixes possible language error caused by wrong lng set in cookie during handle

### v1.6.1
- support for null values in resources
- reuse tests from client script

### v1.6.0
- supportedLngs now used for all detect functions too
- sendMissingTo Option is now working on server too
- option to disable lng lookup from headers
- fixed setting supported lngs via options
- fixed setting options for webtranslate

### v1.5.15
- fixed missing argument in remove route
- updated client

### v1.5.14
- set cookie name
- some small fixes
- set supportedlngs for seo routes
- updated client

### v1.5.13
- disable cookie usage by setting init option useCookie to false
- updated client
- updated webtranslate

### v1.5.12
- ignore routes inside handler

### v1.5.11
- fixed bug for returning objTree in combination with namespace

### v1.5.10
- changable querystring param to look language up from
- yaml backend

### v1.5.9
- bugfix  throws error on savemissing key
- force lng from path (issue 32)

### v1.5.8
- updated client
- updated i18next-webtranslate

### v1.5.7
- bugfix translations won't be in users language in async cases
- bugfix use resSetPath instead of get to save missing resources

### v1.5.6
- bugfix in detectlng

### v1.5.5
- updated client
- detectLng from route: detectLngFromPath index of route path, eg. 0 -> /en-US/myPage.html, 1 -> /cms/en-US/myPage.html
- addRoutes, eg. /en/introduction, /de/einfuehrung,..

### v1.5.0
- remove route added
- optional arg authenticated for postmissing, change, remove routes
- integrated i18next-webtranslate
- couchDb Backend

### v1.4.1
- post processing for translation
- updated client

### v1.4.0
- preload multiple languages
- translate key to other language than current
- updated client to 1.4.0
- added missing tests
- new documentation

### v1.3.4
- updated client
- fixed bug in update route

### v1.3.3

- fixed response type of post routes to json
- updated client

### v1.3.1

- debugging / logging
- updated client
- easier override for backend implementation

### v1.2.5

- added function to grap current language from template (i18n.lng())
- merged [pull request #15](https://github.com/jamuhl/i18next-node/pull/15) for express.js v3 support
- updated client

### v1.2.3 & v1.2.4

- updated client

### v1.2.0

- keep version in sync with clientside version
- better support to override default sync
- support for translation contexts
- fixed init without options, callback

### v0.5.0

- options saveMissing
- multiple plurals
- bug fix

### v0.0.1

- tests with mocha
- multi-namespace support
- loading files from filesystem
- clientscript support
- graceful fallback en-US -> en -> fallbackLng
- support for pluralized strings
- insertion of variables into translations
- translation nestingto translations


src/i18next.helpers.js
function _extend(target, source) {
    if (!source || typeof source === 'function') {
        return target;
    }
    
    for (var attr in source) { target[attr] = source[attr]; }
    return target;
}

function _each(object, callback, args) {
    var name, i = 0,
        length = object.length,
        isObj = length === undefined || typeof object === "function";

    if (args) {
        if (isObj) {
            for (name in object) {
                if (callback.apply(object[name], args) === false) {
                    break;
                }
            }
        } else {
            for ( ; i < length; ) {
                if (callback.apply(object[i++], args) === false) {
                    break;
                }
            }
        }

    // A special, fast, case for the most common use of each
    } else {
        if (isObj) {
            for (name in object) {
                if (callback.call(object[name], name, object[name]) === false) {
                    break;
                }
            }
        } else {
            for ( ; i < length; ) {
                if (callback.call(object[i], i, object[i++]) === false) {
                    break;
                }
            }
        }
    }

    return object;
}

function _ajax(options) {

    // v0.5.0 of https://github.com/goloroden/http.js
    var getXhr = function (callback) {
        // Use the native XHR object if the browser supports it.
        if (window.XMLHttpRequest) {
            return callback(null, new XMLHttpRequest());
        } else if (window.ActiveXObject) {
            // In Internet Explorer check for ActiveX versions of the XHR object.
            try {
                return callback(null, new ActiveXObject("Msxml2.XMLHTTP"));
            } catch (e) {
                return callback(null, new ActiveXObject("Microsoft.XMLHTTP"));
            }
        }

        // If no XHR support was found, throw an error.
        return callback(new Error());
    };

    var encodeUsingUrlEncoding = function (data) {
        if(typeof data === 'string') {
            return data;
        }

        var result = [];
        for(var dataItem in data) {
            if(data.hasOwnProperty(dataItem)) {
                result.push(encodeURIComponent(dataItem) + '=' + encodeURIComponent(data[dataItem]));
            }
        }

        return result.join('&');
    };

    var utf8 = function (text) {
        text = text.replace(/\r\n/g, '\n');
        var result = '';

        for(var i = 0; i < text.length; i++) {
            var c = text.charCodeAt(i);

            if(c < 128) {
                    result += String.fromCharCode(c);
            } else if((c > 127) && (c < 2048)) {
                    result += String.fromCharCode((c >> 6) | 192);
                    result += String.fromCharCode((c & 63) | 128);
            } else {
                    result += String.fromCharCode((c >> 12) | 224);
                    result += String.fromCharCode(((c >> 6) & 63) | 128);
                    result += String.fromCharCode((c & 63) | 128);
            }
        }

        return result;
    };

    var base64 = function (text) {
        var keyStr = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/=';

        text = utf8(text);
        var result = '',
                chr1, chr2, chr3,
                enc1, enc2, enc3, enc4,
                i = 0;

        do {
            chr1 = text.charCodeAt(i++);
            chr2 = text.charCodeAt(i++);
            chr3 = text.charCodeAt(i++);

            enc1 = chr1 >> 2;
            enc2 = ((chr1 & 3) << 4) | (chr2 >> 4);
            enc3 = ((chr2 & 15) << 2) | (chr3 >> 6);
            enc4 = chr3 & 63;

            if(isNaN(chr2)) {
                enc3 = enc4 = 64;
            } else if(isNaN(chr3)) {
                enc4 = 64;
            }

            result +=
                keyStr.charAt(enc1) +
                keyStr.charAt(enc2) +
                keyStr.charAt(enc3) +
                keyStr.charAt(enc4);
            chr1 = chr2 = chr3 = '';
            enc1 = enc2 = enc3 = enc4 = '';
        } while(i < text.length);

        return result;
    };

    var mergeHeaders = function () {
        // Use the first header object as base.
        var result = arguments[0];

        // Iterate through the remaining header objects and add them.
        for(var i = 1; i < arguments.length; i++) {
            var currentHeaders = arguments[i];
            for(var header in currentHeaders) {
                if(currentHeaders.hasOwnProperty(header)) {
                    result[header] = currentHeaders[header];
                }
            }
        }

        // Return the merged headers.
        return result;
    };

    var ajax = function (method, url, options, callback) {
        // Adjust parameters.
        if(typeof options === 'function') {
            callback = options;
            options = {};
        }

        // Set default parameter values.
        options.cache = options.cache || false;
        options.data = options.data || {};
        options.headers = options.headers || {};
        options.jsonp = options.jsonp || false;
        options.async = options.async === undefined ? true : options.async;

        // Merge the various header objects.
        var headers = mergeHeaders({
            'accept': '*/*',
            'content-type': 'application/x-www-form-urlencoded;charset=UTF-8'
        }, ajax.headers, options.headers);

        // Encode the data according to the content-type.
        var payload;
        if (headers['content-type'] === 'application/json') {
            payload = JSON.stringify(options.data);
        } else {
            payload = encodeUsingUrlEncoding(options.data);            
        }

        // Specially prepare GET requests: Setup the query string, handle caching and make a JSONP call
        // if neccessary.
        if(method === 'GET') {
            // Setup the query string.
            var queryString = [];
            if(payload) {
                queryString.push(payload);
                payload = null;
            }

            // Handle caching.
            if(!options.cache) {
                queryString.push('_=' + (new Date()).getTime());
            }

            // If neccessary prepare the query string for a JSONP call.
            if(options.jsonp) {
                queryString.push('callback=' + options.jsonp);
                queryString.push('jsonp=' + options.jsonp);
            }

            // Merge the query string and attach it to the url.
            queryString = queryString.join('&');
            if (queryString.length > 1) {
                if (url.indexOf('?') > -1) {
                    url += '&' + queryString;
                } else {
                    url += '?' + queryString;
                }
            }

            // Make a JSONP call if neccessary.
            if(options.jsonp) {
                var head = document.getElementsByTagName('head')[0];
                var script = document.createElement('script');
                script.type = 'text/javascript';
                script.src = url;
                head.appendChild(script);                
                return;
            }
        }

        // Since we got here, it is no JSONP request, so make a normal XHR request.
        getXhr(function (err, xhr) {
            if(err) return callback(err);

            // Open the request.
            xhr.open(method, url, options.async);

            // Set the request headers.
            for(var header in headers) {
                if(headers.hasOwnProperty(header)) {
                    xhr.setRequestHeader(header, headers[header]);
                }
            }

            // Handle the request events.
            xhr.onreadystatechange = function () {
                if(xhr.readyState === 4) {
                    var data = xhr.responseText || '';

                    // If no callback is given, return.
                    if(!callback) {
                        return;
                    }

                    // Return an object that provides access to the data as text and JSON.
                    callback(xhr.status, {
                        text: function () {
                            return data;
                        },

                        json: function () {
                            return JSON.parse(data);
                        }
                    });
                }
            };

            // Actually send the XHR request.
            xhr.send(payload);
        });
    };

    // Define the external interface.
    var http = {
        authBasic: function (username, password) {
            ajax.headers['Authorization'] = 'Basic ' + base64(username + ':' + password);
        },

        connect: function (url, options, callback) {
            return ajax('CONNECT', url, options, callback);            
        },

        del: function (url, options, callback) {
            return ajax('DELETE', url, options, callback);            
        },

        get: function (url, options, callback) {
            return ajax('GET', url, options, callback);
        },

        head: function (url, options, callback) {
            return ajax('HEAD', url, options, callback);
        },

        headers: function (headers) {
            ajax.headers = headers || {};
        },

        isAllowed: function (url, verb, callback) {
            this.options(url, function (status, data) {
                callback(data.text().indexOf(verb) !== -1);
            });
        },

        options: function (url, options, callback) {
            return ajax('OPTIONS', url, options, callback);
        },

        patch: function (url, options, callback) {
            return ajax('PATCH', url, options, callback);            
        },

        post: function (url, options, callback) {
            return ajax('POST', url, options, callback);            
        },
        
        put: function (url, options, callback) {
            return ajax('PUT', url, options, callback);            
        },

        trace: function (url, options, callback) {
            return ajax('TRACE', url, options, callback);
        }
    };


    var methode = options.type ? options.type.toLowerCase() : 'get';

    http[methode](options.url, options, function (status, data) {
        if (status === 200) {
            options.success(data.json(), status, null);
        } else {
            options.error(data.text(), status, null);
        }
    });
}

var _cookie = {
    create: function(name,value,minutes) {
        var expires;
        if (minutes) {
            var date = new Date();
            date.setTime(date.getTime()+(minutes*60*1000));
            expires = "; expires="+date.toGMTString();
        }
        else expires = "";
        document.cookie = name+"="+value+expires+"; path=/";
    },

    read: function(name) {
        var nameEQ = name + "=";
        var ca = document.cookie.split(';');
        for(var i=0;i < ca.length;i++) {
            var c = ca[i];
            while (c.charAt(0)==' ') c = c.substring(1,c.length);
            if (c.indexOf(nameEQ) === 0) return c.substring(nameEQ.length,c.length);
        }
        return null;
    },

    remove: function(name) {
        this.create(name,"",-1);
    }
};

var cookie_noop = {
    create: function(name,value,minutes) {},
    read: function(name) { return null; },
    remove: function(name) {}
};



// move dependent functions to a container so that
// they can be overriden easier in no jquery environment (node.js)
var f = {
    extend: $ ? $.extend : _extend,
    each: $ ? $.each : _each,
    ajax: $ ? $.ajax : _ajax,
    cookie: typeof document !== 'undefined' ? _cookie : cookie_noop,
    detectLanguage: detectLanguage,
    log: function(str) {
        if (o.debug && typeof console !== "undefined") console.log(str);
    },
    toLanguages: function(lng) {
        var languages = [];
        if (typeof lng === 'string' && lng.indexOf('-') > -1) {
            var parts = lng.split('-');

            lng = o.lowerCaseLng ? 
                parts[0].toLowerCase() +  '-' + parts[1].toLowerCase() :
                parts[0].toLowerCase() +  '-' + parts[1].toUpperCase();

            if (o.load !== 'unspecific') languages.push(lng);
            if (o.load !== 'current') languages.push(parts[0]);
        } else {
            languages.push(lng);
        }

        if (languages.indexOf(o.fallbackLng) === -1 && o.fallbackLng) languages.push(o.fallbackLng);

        return languages;
    },
    regexEscape: function(str) {
        return str.replace(/[\-\[\]\/\{\}\(\)\*\+\?\.\\\^\$\|]/g, "\\$&");
    }
};